### ISSUE TITLE
NPE  in CompletableFuture on textDocument/codeAction failed

### ISSUE THREAD

**Issue by apupier** [redhat-developer/vscode-quarkus#185]
When i have an application.properties file with this kind of content:

`camel.component.ti=plop`

 and that I click inside itthere is an NPE in Quartus Tools output reported:

```
[Error - 2:34:36 PM] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1081)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
	... 6 more
```

the application.properties file is outside of any projects (if it might be the cause)

**Comment by xorye** [comment]
Hi @apupier , do other LSP features like completion or hover work? Do you see completion options? Do they also result in a NPE?

**Comment by apupier** [comment]
other Camel LSP features are working.

Which completion should I expect from quarkus on a standalone file?
If I call completion at the end of this content:
```
camel.component.ti=plop
quarkus.
```

I have no completion.

if I hover over the quakus word, I have this other NPE:

```
[Error - 9:05:03 AM] Request textDocument/hover failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1081)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
	... 6 more
```

**Comment by angelozerr** [comment]
> the application.properties file is outside of any projects (if it might be the cause)

it's the cause thanks @apupier for your relevant information.

@xorye when you open an application.properties which doesn't belong to a project which is not opened in the vscode workspace, the JDT LS extension throw an exception https://github.com/redhat-developer/quarkus-ls/blob/master/microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java#L88 

and when completion future on client side try to get the project info at https://github.com/redhat-developer/quarkus-ls/blob/0313ab1ecdeb1178fbd218c8290b42d6d2b68127/microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/MicroProfileProjectInfoCache.java#L197 the info is null which causes the NPE.

Because info is null:

```java
thenApply(info -> new MicroProfileProjectInfoWrapper(info));
```
I think we should fix this NPE on JDT LS extesnion side and since we could have other LS extension (like IJ LS extension) which could throw errors, we should take care of exception on MicroProfile LS side too.

# On JDT LS extension side 

In https://github.com/redhat-developer/quarkus-ls/blob/master/microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java#L88 it should returns an instance of MicroProfileProjectInfo with empty properties.

In the https://github.com/redhat-developer/vscode-quarkus/issues/45 issue we will improve it to say that the project info is not a mircroprofile project (application.properties doesn't belong to a Microprofile project and in this case validation, hover, etc will be disabled).

# On MicroProfile LS side 

I though `thenApply` was called only if there no exception (perhaps throwing on JDT LS extesnion side `UnsupportedOperationException` is a bad idea?). 

I think we should:

 *  log the error
 * returns an empty project info

In other words, here the idea (not tested):

```java
CompletableFuture<MicroProfileProjectInfo> future = provider.getProjectInfo(params). //
	exceptionally(ex -> {
		// Log error
						
		// return an empty project info						
		return new MicroProfileProjectInfoWrapper(EMPTY_PROJECT_INFO);
	}).thenApply(info -> new MicroProfileProjectInfoWrapper(info));
``` 

And add EMPTY_PROJECT_INFO in the MicroProfileProjectInfoCache:
```java
private static final MicroProfileProjectInfo EMPTY_PROJECT_INFO;
	
	static {
		EMPTY_PROJECT_INFO = new MicroProfileProjectInfo();
		EMPTY_PROJECT_INFO.setProperties(Collections.emptyList());
		EMPTY_PROJECT_INFO.setHints(Collections.emptyList());
		EMPTY_PROJECT_INFO.setProjectURI("");
	}
```
